src/main.c

```diff
@@ -3128,6 +3088,66 @@ static void setup_configurators(void)
     h2o_config_register_status_handler(&conf.globalconf, &extra_status_handler);
 }
 
+static int dup_listener(int listener, int reuseport)
+{
+    int fd = -1;
+#if H2O_USE_REUSEPORT
+    if (reuseport) {
+        int type;
+        socklen_t typelen = sizeof(type);
+        struct sockaddr_storage ss;
+        socklen_t sslen = sizeof(ss);
+        if (getsockopt(listener, SOL_SOCKET, SO_TYPE, &type, &typelen) != 0) {
+            perror("failed to obtain the type of a listening socket");
+            abort();
+        }
+        assert(type == SOCK_DGRAM || type == SOCK_STREAM);
+        if (getsockname(listener, (struct sockaddr *)&ss, &sslen) != 0) {
+            perror("failed to obtain local address of a listening socket");
+            abort();
+        }
+        if ((fd = open_listener(ss.ss_family, type, type == SOCK_STREAM ? IPPROTO_TCP : IPPROTO_UDP, (struct sockaddr *)&ss,
+                                sslen)) != -1) {
+            if (type == SOCK_DGRAM)
+                setsockopt_recvpktinfo(fd, ss.ss_family);
+        } else {
+            perror("failed to bind additional listener");
+            abort();
+        }
+    }
+#endif
+    if (!reuseport && (fd = dup(listener)) == -1) {
+        perror("failed to dup listening socket");
+        abort();
+    }
+    set_cloexec(fd);
+    return fd;
+}
+
+static void create_per_thread_listeners(void)
+{
+    for (size_t i = 0; i != conf.num_listeners; ++i) {
+        struct listener_config_t *listener_config = conf.listeners[i];
+        int reuseport = 0;
+
+#if H2O_USE_REUSEPORT
+        socklen_t reuseportlen = sizeof(reuseport);
+        if (getsockopt(listener_config->fds.entries[0], SOL_SOCKET, H2O_SO_REUSEPORT, &reuseport, &reuseportlen) != 0) {
+            perror("gestockopt(SO_REUSEPORT) failed");
+            abort();
+        }
+        assert(reuseportlen == sizeof(reuseport));
+#endif
```

Comment:
I was thinking of moving this logic (of detecting the source file descriptor has SO_REUSEPORT being set) into `dup_listener` as well, so that `dup_listener` would take only one argument.

I think keeping the detection logic here is fine, but if we are to keep it, I think that we should better not pass the result into `dup_listener` that supports two ways of creating a duplicate file descriptor. Rather, we should write and call a function that creates a duplicate using SO_REUSEPORT, or call `dup`, depending on the value of the `reuseport` flag.

WDYT?

Reply:
That sounds good, i wanted to avoid the extra call, but it's a non-issue. What do you think of d016a0b ?